src/Servers/HttpSys/src/UrlPrefixCollection.cs

```diff
@@ -138,12 +146,56 @@ internal void RegisterAllPrefixes(UrlGroup urlGroup)
             {
                 _urlGroup = urlGroup;
                 // go through the uri list and register for each one of them
-                foreach (var pair in _prefixes)
+                // Call ToList to avoid modification when enumerating.
+                foreach (var pair in _prefixes.ToList())
+                {
+                    var urlPrefix = pair.Value;
+                    if (urlPrefix.PortValue == 0)
+                    {
+                        if (urlPrefix.IsHttps)
+                        {
+                            throw new InvalidOperationException("Cannot bind to port 0 with https.");
+                        }
+
+                        if (!FindHttpPortUnsynchronized(pair.Key, urlPrefix))
+                        {
+                            throw new HttpSysException(Marshal.GetLastWin32Error(), "Could not bind to port 0.");
+                        }
+                    }
+                    else
+                    {
+                        // We'll get this index back on each request and use it to look up the prefix to calculate PathBase.
+                        _urlGroup.RegisterPrefix(pair.Value.FullPrefix, pair.Key);
+                    }
+                }
+            }
+        }
+
+        private bool FindHttpPortUnsynchronized(int key, UrlPrefix urlPrefix)
+        {
+            for (var index = BasePortIndex; index < MaxPortIndex; index++)
+            {
+                try
+                {
+                    // Bit of complicated math to always try 3000 ports, starting from NextPortIndex + 5000,
+                    // circling back around if we go above 8000 back to 5000, and so on.
+                    var port = ((index + NextPortIndex) % MaxPortIndex) + BasePort;
+
+                    Debug.Assert(port >= 5000 || port < 8000);
+
+                    var newPrefix = UrlPrefix.Create(urlPrefix.Scheme, urlPrefix.Host, port, urlPrefix.Path);
+                    _urlGroup.RegisterPrefix(newPrefix.FullPrefix, key);
+                    _prefixes[key] = newPrefix;
+
+                    NextPortIndex = BasePortIndex + 1;
+                    return true;
+                }
+                catch (HttpSysException)
```

Comment:
This should be checking for a specific error code before retrying.

Reply:
I see 3 different potential error code when binding to a port so far:
- ERROR_ACCESS_DENIED
- ERROR_SHARING_VIOLATION
- ERROR_ALREADY_REGISTERED

I could check for all 3 specific error codes, but at some point, I think it's probably better to just retry for any exception.

Reply:
3 is fine to check for. There are many errors that shouldn't be retried such as invalid prefix formats. If this were in test code I'd be less concerned, but this is in the product now.

Reply:
I added checks for them.